### PR TITLE
Custom Edit Actions

### PR DESCRIPTION
**Pull request by cbaltzer**
Allows for adding new actions to the selection menu.

The `EditingAction` enum has be converted to a struct, and the existing cases as static values. This should _hopefully_ not break compatibility.

New actions can be created like so:

```swift
        var config = EPUBNavigatorViewController.Configuration.init()

        let actions = [
            EditingAction(rawValue: "highlight:", name: "Highlight")
        ]
        config.editingActions = actions

        self.navigator = EPUBNavigatorViewController(publication: publication,
                                                     initialLocation: nil,
                                                     resourcesServer: resourceServer,
                                                     config: config)
```

The action's implementation can be added as an extension on `WKWebView`:

```swift
extension WKWebView {
    @objc func highlight(_ sender: UIMenuItem) {
        print("Highlighting some text")
    }
}
```

### CODE REVIEW

**Review by mickael-menu — CHANGES_REQUESTED** [review]
I'm not a fan of the term "editing action" which is a bit confusing IMHO in the selection context, since there's no actual editing being done. So I wouldn't conflate the existing "editing actions" and the new menu items.

Maybe we can keep the editing actions (i.e. native menu items) as-is but add a new struct only for the custom menu items. This would also make the target more explicit than your example extending `WKWebView`. For example:


```swift
public struct MenuItem {
    public let title: String
    
    public init(title: String, target: Any, action: Selector) {
        self.title = title
        self.action = .targetAction(TargetAction(target: target, action: action)))
    }

    public init(title: String, block: @escaping (Any?) -> Void) {
        self.title = title
        self.action = .block(block)
    }

    public invoke(from sender: Any? = nil) {
        switch action {
            case .block(let block):
                block(sender)
            case .targetAction(let targetAction):
                targetAction.invoke(from: sender)
        }
    }

    private let action: Action
    private enum Action {
        case targetAction(TargetAction)
        case block((Any?) -> Void)
    }
}
```

To be used like this:

```swift
        var config = EPUBNavigatorViewController.Configuration()
        // Still available for native actions.
        config.editingActions = [.copy]

        config.selectionMenuItems = [
            MenuItem(title: "Highlight", target: self, action: #selector(highlight:)),
            MenuItem(title: "Annotate") { sender in
                ...
            }
        ]
```

What do you think?

With the 3.0 we will probably unify a bit the Selection API and deprecate "editing actions".

File: r2-navigator-swift/EditingAction.swift [hunk 1]
```diff
@@ -14,30 +14,56 @@ import UIKit
...
     func editingActionsDidPreventCopy(_ editingActions: EditingActionsController)
     
 }
 
 
 /// Handles the authorization and check of editing actions.
-final class EditingActionsController {
-    
-    public weak var delegate: EditingActionsControllerDelegate?
+public final class EditingActionsController {
```
[the remark this answers is not in the document]
Comment (cbaltzer) [reply r2-navigator-swift/EditingAction.swift:62]: Ah, missed that. I thought the selectors would need to be added here instead of the `WKWebView`.

**Comment by mickael-menu** [comment]
@cbaltzer Do you plan to follow-up on this PR? It is quite promising.

**Review by mickael-menu — APPROVED** [review]
@cbaltzer I've reworked your PR and now I see where your implementation was coming from. With `UIMenuController`, there's no way to provide a closure or even (easily) a custom target/selector pair.

So let's embrace the editing actions system of `UIMenuController`. I just fixed a bit your PR to simplify the usage in reading apps. Now you must provide a valid selector for safety, and it just needs to be implemented in any class of the responder chain, most likely the wrapping `UIViewController`, for example:

```swift
class EPUBViewController: UIViewController {
    init(publication: Publication) {
        let navigator = EPUBNavigatorViewController(
            publication: publication,
            config: EPUBNavigatorViewController.Configuration(
                editingActions: [
                    .copy,
                    .share,
                    EditingAction(title: "Highlight", action: #selector(highlight))
                ]
            ))
    }

    @objc func highlight(_ sender: Any) {
        print("Highlighting")
    }
}
```

(Or nicer IMHO)

```swift
extension EditingAction {
    static let highlight = EditingAction(title: "Highlight", action: #selector(BaseReaderViewController.highlight))
}

editingActions: [.copy, .share, .highlight]
```
Next I'll be working on a `SelectableNavigator` protocol to get the information about the current selection, and optionally fully override the `UIMenuController` to be able to provide a custom selection popup.

**Pull request merged**